### PR TITLE
refactor: store user shortcodes as [user=ID]

### PR DESCRIPTION
**Pull request by wescopeland**
**Command to run after deploy:**
```
sail artisan ra:community:migrate-user-shortcodes
```

---

This PR adjusts how `[user]` shortcodes are stored in the `ForumTopicComment` and `messages` tables.

With this PR's changes, a body like:
```
Hello [user=Scott]. I hope you are having a nice day.
```
will now be stored in the DB as:
```
Hello [user=1]. I hope you are having a nice day.
```

This is all done invisibly to the end user. A command is included to migrate the `ForumTopicComment` and `messages` table to the new format:

```
sail artisan ra:community:migrate-user-shortcodes
sail artisan ra:community:migrate-user-shortcodes --undo
```

### CODE REVIEW

File: app/Helpers/shortcode.php [hunk 1]
```diff
@@ -58,11 +59,29 @@ function url2shortcode(string $url): string
     return $link;
 }
 
+function convertUserUrlsToShortcodes(string $input): string
+{
+    // Find all instances of user URLs in the input.
+    preg_match_all("~https?://(?:[\w\-]+\.)?retroachievements\.org/user/([\w\-]+)(/?(?![\w/?]))~i", $input, $matches);
```
Comment (Jamiras) [inline app/Helpers/shortcode.php:65]: localhost user link seems to be handled by the preview functionality, but not this code.
image

File: app/Support/Shortcode/Shortcode.php [hunk 2]
```diff
@@ -40,6 +42,29 @@ public static function render(string $input, array $options = []): string
...
+        // query so someone doesn't inadvertently slam the database.
+        preg_match_all('/\[user=(.*?)\]/', $input, $matches);
+        $usernames = $matches[1];
+
+        if (empty($usernames)) {
+            return $input;
+        }
+
+        // Fetch all users by username in a single query.
+        $users = User::whereIn('User', $usernames)->get()->keyBy('User');
```
Comment (Jamiras) [inline app/Support/Shortcode/Shortcode.php:57]: Soft-deleted users aren't being converted, but appear to be functional.
```
| Ask [user=Ansku].         |
```
Also doesn't seem to be converting incorrectly cased users:
```
| Ask [user=14863].         |
| Ask [user=Televandalist]. |
```

File: app/Helpers/shortcode.php (hunk 1 above)
Comment (wescopeland) [reply app/Helpers/shortcode.php:65]: Swapping the execution order of `Shortcode::convertUserShortcodesToUseIds()` and `normalize_shortcodes()` seems to resolve this issue. Updated in latest.

File: app/Support/Shortcode/Shortcode.php (hunk 2 above)
Comment (wescopeland) [reply app/Support/Shortcode/Shortcode.php]: Good catches - both fixed in latest.

Comment (Jamiras) [reply app/Support/Shortcode/Shortcode.php]: Deleted user is being converted to an int, but not converted back.
image

File: app/Helpers/shortcode.php (hunk 1 above)
Comment (Jamiras) [reply app/Helpers/shortcode.php:65]: Seems to be working in forums posts, but not messages.
image

Comment (wescopeland) [reply app/Helpers/shortcode.php:65]: Should be fixed in latest. Looks like this issue also happens in prod?

File: app/Support/Shortcode/Shortcode.php (hunk 2 above)
Comment (wescopeland) [reply app/Support/Shortcode/Shortcode.php]: Resolved in latest.

File: app/Helpers/shortcode.php (hunk 1 above)
Comment (Jamiras) [reply app/Helpers/shortcode.php:65]: Verified the URL->[user] functionality is not present in master. The delete user is handled correctly there.

Everything seems to be working correctly on the branch now.

**Review by Jamiras — APPROVED** [review]

**Pull request merged**